rust/catalyst-types/src/id_uri/role_index.rs

```diff
@@ -12,54 +12,82 @@ use thiserror::Error;
 /// Role Index parsing error
 #[derive(Display, Error, Debug)]
 #[allow(clippy::module_name_repetitions)]
-pub enum RoleIndexError {
+pub enum RoleIdError {
+    /// Invalid parse
+    Parse(#[from] ParseIntError),
+
     /// Invalid Role Index
-    InvalidRole(#[from] ParseIntError),
+    InvalidRole(u8),
 }
 
 /// Project Catalyst User Role Index.
 ///
 /// <https://github.com/input-output-hk/catalyst-CIPs/blob/x509-catalyst-role-definitions/CIP-XXXX/README.md>
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RoleIndex(u16);
-
-impl RoleIndex {
-    /// Delegated Representative
-    pub const DREP: RoleIndex = RoleIndex(1);
-    /// Proposer
-    pub const PROPOSER: RoleIndex = RoleIndex(3);
-    /// Default Role 0
-    pub const ROLE_0: RoleIndex = RoleIndex(0);
+#[repr(u8)]
+#[non_exhaustive]
+pub enum RoleId {
+    /// Can Vote and Comment on Proposals.
+    Role0 = 0,
+    /// Allows for voters to give their voting power to the dRep who then votes on their
+    /// behalf.
+    DelegatedRepresentative = 1,
+    /// Assigns a voters voting power to a registered dRep.
+    VoterDelegation = 2,
+    /// Allows the Propers signing key to also be used as a public encryption key.
+    Proposer = 3,
+}
 
-    /// Is the `RoleIndex` the default value (Role 0)
+impl RoleId {
+    /// Is the `RoleId` the default value [`Self::Role0`]
     #[must_use]
     pub fn is_default(self) -> bool {
-        self == Self::ROLE_0
+        self == Self::Role0
     }
 }
 
-impl Default for RoleIndex {
+impl Default for RoleId {
     fn default() -> Self {
-        Self::ROLE_0
+        Self::Role0
     }
 }
 
-impl From<u16> for RoleIndex {
-    fn from(value: u16) -> Self {
-        Self(value)
+impl TryFrom<u8> for RoleId {
+    type Error = RoleIdError;
+
+    fn try_from(value: u8) -> Result<Self, RoleIdError> {
+        match value {
+            0 => Ok(Self::Role0),
+            1 => Ok(Self::DelegatedRepresentative),
+            2 => Ok(Self::VoterDelegation),
+            3 => Ok(Self::Proposer),
+            _ => Err(RoleIdError::InvalidRole(value)),
+        }
     }
 }
 
-impl FromStr for RoleIndex {
-    type Err = RoleIndexError;
+impl FromStr for RoleId {
```

Comment:
Try using https://docs.rs/strum_macros/latest/strum_macros/derive.EnumString.html instead.

Reply:
Lets maybe do the following:
1. Try to remove this implementation `impl FromStr for RoleId` and look where it is used (also check on the cat-gateway side)
2. If there are few places where it affects so lets stick with the `strum`
3. If it would be a huge breaking change, lets stick with the original implementation `RoleId::from_str("0")`

Reply:
Yes, I see. Lets keep original implementation `RoleId::from_str("0")`, we cannot change the uri format, it cannot be `role0` inside the catalyst id